### PR TITLE
Customizable inputs/outputs and modular data processing code

### PR DESCRIPTION
**Pull request by ftavella**
This PR moves the core of the data processing step out of the model functions into the `DataProcessor` class. 

To handle multiple inputs, I created two classes `ModelInput1D` and `ModelInputSpectrogram` which encompass the type of inputs we will be dealing with and makes it easier to expand functionality to other types of inputs. In addition, the `PSGType` class allows to specify if the PSG used has N4 stage or not.

To handle multiple outputs, I created the `ModelOutputType` class with helps determine if the prediction will be done for sleep/wake or wake/light/deep/rem.

For testing these functions I added a `mock_data_sets` folder with mock data created with a function inside the data sets notebook (note that this function is not exported to the library).

In addition, this PR solves #15, where we were creating a spectrogram using the time dimension rather than the `x` variable.

Both `SGDLogisticRegression` and Mads Olsen's model have been updated to use this new functionality

### CODE REVIEW

File: pisces/data_sets.py
Comment (ericcanton) [inline pisces/data_sets.py:396]: cut this comment?

File: pisces/data_sets.py
Comment (ericcanton) [inline pisces/data_sets.py:578]: I wonder if `.NO_N4` is a better default value because it has been discontinued in the literature.

File: pisces/models.py
Comment (ericcanton) [inline pisces/models.py:110]: FWIW, parsing the logging messages of another package (or any logging messages) as part of your code is generally considered very fragile, because it requires the strings to be basically identical across subversions. 

But if the scikit-learn maintainers fix a typo in a string, or add a newline, that can wreck any downstream string parsing. From their perspective, though, it is kind of a trivial detail that may only bump the patch version because (hence should be compatible with any downstream code). 

That being said, I do like the idea of capturing the losses here in the way we can do elsewhere. I'm not seeing any built-in capabilities of sklearn, so maybe output parsing is as good as it gets for now. But in that case, we should wrap this parsing in a try-except block in case it fails.

File: pisces/data_sets.py
Comment (ericcanton) [inline pisces/data_sets.py:594]: I think there are 2 nice getters that could simplify some other code

```python
@property
def is_1D(self):
    return isinstance(self.model_input, ModelInput1D)

@property
def is_specgtrogram(self):
    isinstance(model_input, ModelInputSpectrogram)
```
then we can change the end of this to 
```python
if self.is_1D:
...
elif self.is_spectrogram:
...```
This makes the code more readable and less error-prone elsewhere, like in `preprocess_data_for_subject` later on.

File: pisces/data_sets.py
Comment (ericcanton) [inline pisces/data_sets.py:629]: There is an issue here when `feature_values` is not 1d. You must interpolate each axis separately, as far as I know.

For example:

```python
import numpy as np

t = np.linspace(0, 2*np.pi, 100)
x = np.zeros((len(t), 2))
x[:, 0] = np.sin(t)
x[:, 1] = np.cos(t)

t2 = np.linspace(0, 2*np.pi, 50)
x2 = np.interp(t2, t, x)
```
outputs
```markdown
{
	"name": "ValueError",
	"message": "object too deep for desired array",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[3], line 8
      4 x[:, 1] = np.cos(t)
      6 t2 = np.linspace(0, 2*np.pi, 50)
----> 8 x2 = np.interp(t2, t, x)

File ~/Engineering/Work/pisces/.pisces/lib/python3.11/site-packages/numpy/lib/function_base.py:1599, in interp(x, xp, fp, left, right, period)
   1596     xp = np.concatenate((xp[-1:]-period, xp, xp[0:1]+period))
   1597     fp = np.concatenate((fp[-1:], fp, fp[0:1]))
-> 1599 return interp_func(x, xp, fp, left, right)

ValueError: object too deep for desired array"
}
```

File: pisces/data_sets.py
Comment (ericcanton) [inline pisces/data_sets.py:647]: `self. find_overlapping_time_section(...)` here once making the other change

File: pisces/data_sets.py
Comment (ericcanton) [inline pisces/data_sets.py:737]: `self.find_overlapping_time_section` here

File: pisces/data_sets.py
Comment (ericcanton) [inline pisces/data_sets.py:638]: There's also a problem here for multidimensional vectors, this reshape will be weird (though kind of interesting on it's own...)

**Review by ericcanton — CHANGES_REQUESTED** [review]
This is great work. Just a few places where I see an edge case we need to watch out for.

File: pisces/data_sets.py
Comment (ftavella) [reply pisces/data_sets.py:567]: Yep, you're right

File: pisces/models.py
Comment (ftavella) [reply pisces/models.py:111]: Good call! I was surprised that there isn't a way to obtain the loss history as an output. I think this works for now and definitely the try/except block makes it more robust!

File: pisces/data_sets.py
Comment (ftavella) [reply pisces/data_sets.py:626]: You're right, I updated the code on `get_1D_X_y` to go through all the columns in a feature and feed them one by one to this function. Do you think that's a good idea?

File: pisces/data_sets.py
Comment (ftavella) [reply pisces/data_sets.py:635]: Since we are passing one column at a time, will this be ever be a problem we run into?

**Review by ericcanton — APPROVED** [review]
This looks great. I just want to try it on my computer before we merge.

**Pull request merged**
